File: src/day_15.rs
use std::collections::{HashMap, VecDeque};
// walls = '#'
// cavern = '.'
// goblin = 'G'
// elf = 'E'

struct Combat {}
impl Combat {
    fn round() {
        /*
        1. Identify all ENEMY TARGETS
        2. Identify all valid open squares ('.')
            -> Identify shortest path to enemy and identify the next square it should occupy.
        unit_takes_turn
            -> if not in RANGE {
                |-> Tries to MOVE into range of enemy
               } else {
                |-> ATTACK
               }
        * units only move DIAGNALLY AND HORIZONTALLY

        */
    }
}

struct Board {
    cells: Vec<Vec<char>>,
    cell_neighbours: HashMap<(usize, usize), Vec<(usize, usize)>>,
}

impl Board {
    fn new(input: &str) -> Board {
        let cells = Board::populate_cells(input);
        let cell_neighbours = Board::calculate_paths(&cells);
        Board{cells, cell_neighbours }
    }

    fn populate_cells(input: &str) -> Vec<Vec<char>> {
        let mut cells = Vec::new();
        for line in input.lines() {
            let cols: Vec<char> = line.chars().collect();
            cells.push(cols);
        }

        cells
    }

    fn calculate_paths(cells: &Vec<Vec<char>>) -> HashMap<(usize, usize), Vec<(usize, usize)>> {
        let mut hmap: HashMap<(usize, usize), Vec<(usize, usize)>> = HashMap::new();

        for x in 0..cells.len() {
            for y in 0..cells[x].len() {
                // find valid children
                let mut collection = Vec::new();

                let mut neighbours = Vec::new();

                // up first then left then right, then down
                if x > 0 { neighbours.push((x-1, y)) }
                if y > 0 { neighbours.push((x, y-1)) };
                if x < cells.len()-1 { neighbours.push((x + 1, y)) };
                if y < cells[x].len()-1 { neighbours.push((x, y+1)) };

                for (i, j) in neighbours.into_iter() {
                    match cells[i][j] {
                        '.' | 'G' | 'E' => collection.push((i, j)),
                        _ => (),
                    }
                }

                hmap.entry((x, y)).or_insert(collection);
            }
        }

        hmap
    }

    fn calculate_shortest_path(&mut self, from: (usize, usize), to: (usize, usize)) {
        // First attempt -- only showed me if the path was possible -- but not keeping
        // track of the actual shorest path.

        let mut visited = Vec::new();
        let mut frontier = VecDeque::new();
        frontier.insert(0, from);

        while !frontier.is_empty() {
            let current = frontier.pop_front().unwrap();

            if current == to {
                println!("Found A Path to Destination! => {:?}", visited);
                break
            }
            else {
                if !visited.contains(&current) {
                    println!("Checking neighbours... ");
                    frontier.extend(self.cell_neighbours.get(&current).unwrap().to_vec());
                }
            }

            visited.push(current);
        }

        println!("Unable to find a path to destination.");
    }

    fn calculate_shortest_path2(&mut self, from: (usize, usize), to: (usize, usize)) -> HashMap<(usize, usize), (usize, usize)> {
        // keeping track of the parents which I can use to rebuild the path --- but have not
        // proven to myself that it is the shortest path yet.

        let mut frontier = VecDeque::new();
        frontier.insert(0, from);

        let mut came_from: HashMap<(usize, usize), (usize, usize)> = HashMap::new();
        came_from.entry(from).or_insert((999, 999));

        while !frontier.is_empty() {
            let current = frontier.pop_front().unwrap();
            for next in self.cell_neighbours.get(&current).unwrap().to_vec() {
                if !came_from.contains_key(&next) {
                    frontier.push_back(next);
                    came_from.entry(next).or_insert(current);
                }
            }
        }
        return came_from;
    }

    fn display(&self) {
        for x in 0..self.cells.len() {
            let mut buffer = String::new();
            for y in 0..self.cells[x].len() {
                buffer.push(self.cells[x][y]);
                println!("({},{})", x, y);
            }
            println!("{}", buffer);
        }
    }
}

#[test]
fn test_scan_actors_in_read_order() {
    let path = "data/day_15_test_scan.txt";
    let data = ::read(path);
    let mut board = Board::new(&data);
    board.display();

    // currenty just confirms there is a connection...
    let mut buffer = Vec::new();
    let to = (1, 5);
    let results: HashMap<(usize, usize), (usize, usize)> = board.calculate_shortest_path2((4,1), to);
    let mut current = &to;
    while !(current == &(999, 999) || !results.contains_key(&current)) {
        buffer.push(current.clone());
        current = &results[&current];
    }
    println!("Shortest Path: {:?}", buffer);
}

File: python/day_15.py
from __future__ import annotations

from collections import defaultdict
from typing import Tuple, List, Set, Iterator, Dict, Optional

import helpers
from pathlib import Path

Pos = Tuple[int, int]


class Node:
    def __init__(self, state, parent: Optional[Node], depth: int) -> None:
        self.state = state
        self.parent = parent
        self.depth = depth


class Unit:
    def __init__(self, pos: Pos, unit_type: str) -> None:
        self.pos = pos
        self.type = unit_type
        self.hp = 200
        self.ap = 3

    def __lt__(self, other: Unit) -> bool:
        return self.pos < other.pos

    def __str__(self) -> str:
        return f'Unit({self.pos}, {self.type}, {self.hp}'


def _neighbours(pos: Pos, blockers: Set[Pos]) -> Iterator[Pos]:
    # print("\t\tOriginal Pos:", pos)
    for n in read_positions(pos):
        # print("\t\t\tTrying:", n, "Is Blocked:", n in blockers)
        if n not in blockers:
            yield n
    # yield pos


def bfs(start: Pos, goal: Pos, units, walls) -> Optional[Node]:
    blockers = {u.pos for u in units if u.hp > 0 and u.pos != goal}.union(walls)
    q: List[Node] = [Node(start, None, 0)]
    visited: Set[Pos] = {start, }
    while q:
        node = q.pop(0)
        if node.state == goal:
            return node

        for neighbour in _neighbours(node.state, blockers):
            if neighbour not in visited:
                visited.add(neighbour)
                q.append(Node(neighbour, node, node.depth + 1))
    return None


def parse(path: Path) -> Tuple[Set[Pos], Set[Unit], List[List[str]]]:
    board: List[List[str]] = [list(line) for line in helpers.lines(path)]
    walls: Set[Pos] = set()
    units: Set[Unit] = set()
    for y in range(len(board)):
        for x in range(len(board[0])):
            char = board[y][x]
            if char in 'GE':
                units.add(Unit((y, x), char))
                board[y][x] = '.'
            elif char in '#':
                walls.add((y, x))

    return walls, units, board


def pprint(round: int, board: List[List[str]], units: Set[Unit]) -> None:
    """Print out a debug board to review each round's results."""
    print(f"Round: {round:02}: Score: {sum(u.hp for u in units if u.hp>0 and u.type == 'G')}")
    for y in range(len(board)):
        scores: List[str] = []
        for x in range(len(board[0])):
            is_unit = False
            for u in units:
                if u.hp > 0 and u.pos == (y, x):
                    is_unit = True
                    print(u.type, end='')
                    scores.append(f'{u.type}({u.hp})')
            if not is_unit:
                print(board[y][x], end='')

        print(' ', ', '.join(scores))


def read_positions(pos: Pos) -> List[Pos]:
    """Read order is like reading an english textbook from top left to bottom right.  There are
    only four possible positions that can surround an Unit and is guaranteed to be in the sorted
    position of first position, then second.  Or Up, left, right, down."""
    u = pos[0] - 1, pos[1]
    d = pos[0] + 1, pos[1]
    l = pos[0], pos[1] - 1
    r = pos[0], pos[1] + 1

    return [u, l, r, d]


def is_beside(unit: Unit, enemies: List[Unit]) -> bool:
    for e in enemies:
        if e.hp > 0 and e.pos in read_positions(unit.pos):
            return True
    return False


def attack(unit: Unit, enemies: List[Unit]) -> None:
    targets = [e for e in enemies if e.hp > 0 and e.pos in read_positions(unit.pos)]
    assert len(targets) <= 4, "Impossible to have more than four enemy positions surrounding a Unit."

    cheapest_unit = None
    cheapest_cost = 201
    for t in targets:
        if t.hp < cheapest_cost:
            cheapest_cost = t.hp
            cheapest_unit = t

    if cheapest_unit is not None:
        cheapest_unit.hp -= unit.ap


def part01(walls, units, board) -> Optional[int]:
    for round in range(1_000):
        # pprint(round, board, units)
        units = {u for u in units if u.hp > 0}

        for unit in sorted(units):
            # Skip over units that are 'dead': hp of zero or less
            if unit.hp <= 0:
                continue
            elves = sorted([u for u in units if u.type == 'E' and u.hp > 0])
            goblins = sorted([u for u in units if u.type == 'G' and u.hp > 0])
            enemies = elves if unit.type == 'G' else goblins

            # End if no enemies to fight
            if len(enemies) == 0:
                # pprint(round, board, units)
                s = sum(u.hp for u in units if u.hp > 0)
                # print('Found on round: ', round, 'with a sum of', s, '=', round * s, 'for', unit)
                return round * s

            # if one step away from any enemy find the enemy that has the lowest HP in reading order and attack
            if not is_beside(unit, enemies):
                # Else, find the shortest path to all enemies and make a move...
                possibles: Dict[int, List[Tuple[Pos, Pos]]] = defaultdict(list)
                blockers = {u.pos for u in units if u.hp > 0}.union(walls)
                start_steps = list(_neighbours(unit.pos, blockers))

                for enemy in enemies:
                    for neighbour in _neighbours(enemy.pos, blockers):
                        for start_step in start_steps:
                            r = bfs(start_step, neighbour, units, walls)
                            if r is not None:
                                #         TargetD          Enemy sq.   Unit.sq
                                possibles[r.depth].append((enemy.pos, start_step))

                if possibles:
                    # Take te lowest key's enemy, unit
                    k, v = sorted(possibles.items())[0]

                    move = sorted(v, key=lambda x: x[0])[0][1]
                    if move not in blockers:
                        unit.pos = move

            attack(unit, enemies)

    return None


def run() -> None:
    path = Path(__file__).parent / '..' / 'data' / 'day_15.txt'
    walls, units, board = parse(path)

    p1 = part01(walls, units, board)
    assert p1 == 227290


if __name__ == "__main__":
    run()
